### PR TITLE
feat: add download, upload, ping tests duration

### PR DESCRIPTION
**Pull request by bluemanos**
In this PR I added a ping, download and upload duration. Plus one common function to measure all the things at once (one by one of course).

### CODE REVIEW

**Comment by r3inbowari** [comment]
Thanks! @bluemanos, should we include the time consuming for fetching user infomation and servers.

**Comment by bluemanos** [comment]
@r3inbowari I can add it, no problem. Indeed these operations also are taking some time, not directly in testing the connection, but in preparation for testing.

My intention was to just get information on how much time the testing took.

**Comment by r3inbowari** [comment]
Hi, @bluemanos, I left some comments, Thank you!

**Comment by bluemanos** [comment]
@r3inbowari where are the comments? I think you start a review, but in the end, did not send them. It happens :D

**Comment by r3inbowari** [comment]
> @r3inbowari where are the comments? I think you start a review, but in the end, did not send them. It happens :D

Sorry!

File: speedtest/request.go [hunk 1]
```diff
@@ -176,10 +185,31 @@ func (s *Server) PingTestContext(ctx context.Context, callback func(latency time
...
+	if err != nil {
+		return err
+	}
+	err = s.UploadTest()
+	if err != nil {
+		return err
+	}
+
 	return nil
 }
```
[the remark this answers is not in the document]
Comment (bluemanos) [reply speedtest/request.go:214]: To run all tests at once rather than run it manually one by one :P

[the remark this answers is not in the document]
Comment (r3inbowari) [reply speedtest/request.go:214]: If it's just for testing, why not hide the method.

[the remark this answers is not in the document]
Comment (bluemanos) [reply speedtest/request.go:214]: Not for testing itself, but to run all speed tests at once. When in my code I want to get a ping latency, download/upload speed than its easily to run just one method :)

File: speedtest/server.go [hunk 2]
```diff
@@ -30,26 +30,34 @@ const (
...
+	TestDuration TestDuration  `json:"test_duration"`
 
 	Context *Speedtest
 }
 
+type TestDuration struct {
+	Ping     *time.Duration `json:"ping"`
+	Download *time.Duration `json:"download"`
+	Upload   *time.Duration `json:"upload"`
+	Total    *time.Duration `json:"total"`
```
[the remark this answers is not in the document]
Comment (r3inbowari) [reply speedtest/server.go:58]:
```suggestion
	Latency      time.Duration `json:"latency"`
	MaxLatency   time.Duration `json:"max_latency"`
	MinLatency   time.Duration `json:"min_latency"`
	Jitter       time.Duration `json:"jitter"`
```

But here, `zero` means no run test.

File: speedtest/request_test.go [hunk 3]
```diff
@@ -59,6 +61,9 @@ func TestUploadTestContext(t *testing.T) {
 	if server.ULSpeed < idealSpeed*(1-delta) || idealSpeed*(1+delta) < server.ULSpeed {
 		t.Errorf("got unexpected server.ULSpeed '%v', expected between %v and %v", server.ULSpeed, idealSpeed*(1-delta), idealSpeed*(1+delta))
 	}
+	if server.TestDuration.Upload == nil || *server.TestDuration.Upload != *server.TestDuration.Total {
+		t.Errorf("cant count test duration, server.TestDuration.Upload=%v, server.TestDuration.Total=%v", server.TestDuration.Upload, server.TestDuration.Total)
```
Comment (r3inbowari) [inline speedtest/request_test.go:65]:
```suggestion
		t.Errorf("can't count test duration, server.TestDuration.Upload=%v, server.TestDuration.Total=%v", server.TestDuration.Upload, server.TestDuration.Total)
```

File: speedtest/request.go [hunk 4]
```diff
@@ -176,10 +185,31 @@ func (s *Server) PingTestContext(ctx context.Context, callback func(latency time
...
+	err = s.DownloadTest()
+	if err != nil {
+		return err
+	}
+	err = s.UploadTest()
+	if err != nil {
+		return err
+	}
+
 	return nil
```
Comment (r3inbowari) [inline speedtest/request.go:213]:
```suggestion
	return s.UploadTest()
```

File: speedtest/server_test.go [hunk 5]
```diff
@@ -115,3 +118,28 @@ func TestCustomServer(t *testing.T) {
...
+	server.testDurationTotalCount()
+	if server.TestDuration.Total.Nanoseconds() != 20001209417 {
+		t.Error("addition in testDurationTotalCount do not works")
+	}
+
+	pingTime := time.Duration(2183156458)
+	server.TestDuration.Ping = &pingTime
+	server.testDurationTotalCount()
+	if server.TestDuration.Total.Nanoseconds() != 22184365875 {
+		t.Error("addition in testDurationTotalCount do not works")
```
Comment (r3inbowari) [inline speedtest/server_test.go:143]:
```suggestion
		t.Error("addition in testDurationTotalCount didn't work")
```

File: speedtest/server_test.go [hunk 6]
```diff
@@ -115,3 +118,28 @@ func TestCustomServer(t *testing.T) {
...
+	server.testDurationTotalCount()
+	if server.TestDuration.Total.Nanoseconds() != 10000805542 {
+		t.Error("addition in testDurationTotalCount do not works")
+	}
+
+	downloadTime := time.Duration(10000403875)
+	server.TestDuration.Download = &downloadTime
+	server.testDurationTotalCount()
+	if server.TestDuration.Total.Nanoseconds() != 20001209417 {
+		t.Error("addition in testDurationTotalCount do not works")
```
Comment (r3inbowari) [inline speedtest/server_test.go:136]:
```suggestion
		t.Error("addition in testDurationTotalCount do didn't work")
```

File: speedtest/server_test.go [hunk 7]
```diff
@@ -115,3 +118,28 @@ func TestCustomServer(t *testing.T) {
...
 }
+
+func TestTotalDurationCount(t *testing.T) {
+	server, _ := CustomServer("https://example.com/upload.php")
+
+	uploadTime := time.Duration(10000805542)
+	server.TestDuration.Upload = &uploadTime
+	server.testDurationTotalCount()
+	if server.TestDuration.Total.Nanoseconds() != 10000805542 {
+		t.Error("addition in testDurationTotalCount do not works")
```
Comment (r3inbowari) [inline speedtest/server_test.go:129]:
```suggestion
		t.Error("addition in testDurationTotalCount didn't work")
```

**Comment by r3inbowari** [comment]
@bluemanos Thanks you!

**Pull request merged**